src/codegen/llvm/llvm_ir_builder.cpp

```diff
@@ -228,11 +228,35 @@ void IRBuilder::set_loop_metadata(llvm::BranchInst* branch) {
 /*                             LLVM instruction utilities                               */
 /****************************************************************************************/
 
+llvm::Value* IRBuilder::create_alloca(const std::string& name, llvm::Type* type) {
+    // If insertion point for `alloca` instructions is not set, then set the first processed
+    // instruction to be the insertion point.
+    if (!alloca_ip) {
+        llvm::Value* alloca = builder.CreateAlloca(type, /*ArraySize=*/nullptr, name);
+        alloca_ip = llvm::cast<llvm::AllocaInst>(alloca);
+        return alloca;
+    }
```

Comment:
so we are assuming that there is always an `alloca` outside loop body and that will be our insertion point, right?
this is fine given that how our kernels look like (`node_id` and other variables are created at the beginning). May be we can add a comment here?

Reply:
There is always one in the entry block since we create `alloca`s for every function argument.  I will add a comment saying that.

For our kernels, allocations will be also pulled out together (before they were separated by the vector for loop:
```llvm
define void @nrn_state_test(%test__instance_var__type* noalias nocapture readonly %mech1) #0 {
  %mech = alloca %test__instance_var__type*, align 8
  %id = alloca i32, align 4
  %node_id = alloca <4 x i32>, align 16
  %v = alloca <4 x double>, align 32
  ; now epilogue variables are also here!
  %epilogue_node_id = alloca i32, align 4
  %epilogue_v = alloca double, align 8
  store %test__instance_var__type* %mech1, %test__instance_var__type** %mech, align 8
  store i32 0, i32* %id, align 4
  ; rest of the kernel
```

Reply:
@pramodk Actually, if we have a void function with no arguments (not real, but still) - this would generate `alloca`s in-place. So I specified explicitly that we need to get the entry block and insert there.

Reply:
looks good. thanks!